File: packages/trust-receipts/src/crypto.ts
/**
 * Cryptographic utilities for Trust Receipts
 *
 * Ed25519 signatures with Node.js crypto sha512.
 * Uses @noble/ed25519 v3 for the actual cryptographic operations.
 */

import { createHash } from 'crypto';

let ed25519Promise: Promise<any> | null = null;

/**
 * Load and configure @noble/ed25519 v3
 */
async function loadEd25519(): Promise<any> {
  if (!ed25519Promise) {
    ed25519Promise = (new Function('return import("@noble/ed25519")')() as Promise<any>).then(
      (ed25519) => {
        // Configure sha512 for @noble/ed25519 v3
        ed25519.hashes.sha512 = (message: Uint8Array) =>
          new Uint8Array(createHash('sha512').update(message).digest());
        return ed25519;
      }
    );
  }
  return ed25519Promise;
}

/**
 * Generate a new Ed25519 key pair
 *
 * @returns Object with privateKey and publicKey as Uint8Array (32 bytes each)
 *
 * @example
 * ```typescript
 * const { privateKey, publicKey } = await generateKeyPair();
 * console.log('Private key:', Buffer.from(privateKey).toString('hex'));
 * console.log('Public key:', Buffer.from(publicKey).toString('hex'));
 * ```
 */
export async function generateKeyPair(): Promise<{
  privateKey: Uint8Array;
  publicKey: Uint8Array;
}> {
  const ed25519 = await loadEd25519();
  const privateKey = (ed25519.utils.randomPrivateKey ?? ed25519.utils.randomSecretKey)();
  const publicKey = await ed25519.getPublicKey(privateKey);
  return { privateKey, publicKey };
}

/**
 * Sign a message with Ed25519
 *
 * @param message - The message to sign (Uint8Array)
 * @param privateKey - The 32-byte Ed25519 private key
 * @returns The 64-byte signature
 */
export async function sign(message: Uint8Array, privateKey: Uint8Array): Promise<Uint8Array> {
  const ed25519 = await loadEd25519();
  return ed25519.sign(message, privateKey);
}

/**
 * Verify an Ed25519 signature
 *
 * @param signature - The 64-byte signature to verify
 * @param message - The original message
 * @param publicKey - The 32-byte Ed25519 public key
 * @returns true if valid, false otherwise
 */
export async function verify(
  signature: Uint8Array,
  message: Uint8Array,
  publicKey: Uint8Array
): Promise<boolean> {
  const ed25519 = await loadEd25519();
  return ed25519.verify(signature, message, publicKey);
}

/**
 * Derive public key from private key
 *
 * @param privateKey - The 32-byte Ed25519 private key
 * @returns The corresponding 32-byte public key
 */
export async function getPublicKey(privateKey: Uint8Array): Promise<Uint8Array> {
  const ed25519 = await loadEd25519();
  return ed25519.getPublicKey(privateKey);
}

/**
 * Compute SHA-256 hash
 *
 * @param data - Data to hash (string or Uint8Array)
 * @returns Hex-encoded hash string
 */
export function sha256(data: string | Uint8Array): string {
  const buffer = typeof data === 'string' ? Buffer.from(data, 'utf8') : data;
  return createHash('sha256').update(buffer).digest('hex');
}

/**
 * Encode bytes to hex string
 */
export function bytesToHex(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString('hex');
}

/**
 * Decode hex string to bytes
 */
export function hexToBytes(hex: string): Uint8Array {
  return new Uint8Array(Buffer.from(hex, 'hex'));
}

File: packages/trust-receipts-python/sonate/crypto.py
"""
Cryptographic utilities for SONATE Trust Receipts

Provides Ed25519 signing/verification, SHA-256 hashing,
and RFC 8785 JSON canonicalization.
"""

import hashlib
from typing import Tuple, Union
from nacl import signing, utils
import json_canonicalize


def sha256(data: Union[str, bytes]) -> str:
    """SHA-256 hash of data"""
    if isinstance(data, str):
        data = data.encode('utf-8')
    return hashlib.sha256(data).hexdigest()


def canonicalize_json(obj: any) -> str:
    """RFC 8785 JSON canonicalization"""
    return json_canonicalize.canonicalize(obj)


def canonicalize_and_hash(obj: any) -> str:
    """Canonicalize object and hash with SHA-256"""
    canonical = canonicalize_json(obj)
    return sha256(canonical)


def generate_key_pair() -> Tuple[str, str]:
    """Generate new Ed25519 key pair"""
    private_key = signing.SigningKey.generate()
    private_hex = private_key.encode(encoder=utils.HexEncoder).decode('ascii')
    public_hex = private_key.verify_key.encode(encoder=utils.HexEncoder).decode('ascii')
    return private_hex, public_hex


def get_public_key(private_key: Union[str, bytes]) -> str:
    """Derive public key from private key"""
    if isinstance(private_key, str):
        private_key = bytes.fromhex(private_key)
    
    signing_key = signing.SigningKey(private_key)
    return signing_key.verify_key.encode(encoder=utils.HexEncoder).decode('ascii')


def sign(message: Union[str, bytes], private_key: Union[str, bytes]) -> str:
    """Sign message with Ed25519 private key"""
    if isinstance(message, str):
        message = message.encode('utf-8')
    if isinstance(private_key, str):
        private_key = bytes.fromhex(private_key)
    
    signing_key = signing.SigningKey(private_key)
    signed_msg = signing_key.sign(message)
    return signed_msg.signature.hex()


def verify(signature: Union[str, bytes], message: Union[str, bytes], public_key: Union[str, bytes]) -> bool:
    """Verify Ed25519 signature"""
    if isinstance(signature, str):
        signature = bytes.fromhex(signature)
    if isinstance(message, str):
        message = message.encode('utf-8')
    if isinstance(public_key, str):
        public_key = bytes.fromhex(public_key)
    
    try:
        verify_key = signing.VerifyKey(public_key)
        verify_key.verify(message, signature)
        return True
    except Exception:
        return False


def hex_to_bytes(hex_string: str) -> bytes:
    """Convert hex string to bytes"""
    return bytes.fromhex(hex_string)


def bytes_to_hex(data: bytes) -> str:
    """Convert bytes to hex string"""
    return data.hex()
